ebean-datasource/src/main/java/io/ebean/datasource/pool/PooledConnection.java

```diff
@@ -118,12 +120,14 @@ final class PooledConnection extends ConnectionDelegator {
     this.pool = pool;
     this.connection = connection;
     this.name = pool.name() + uniqueId;
+    this.originalSchema = pool.schema();
+    this.originalCatalog = pool.catalog();
+    this.initialisedSchema = originalSchema != null;
+    this.initialisedCatalog = originalCatalog != null;
```

Comment:
> So the current **_schema of the connection would be S1, but currentSchema is still null_**. So we prepare a statement for S1, but the cache key contains null

> If we then invoke **_setSchema(null)_** the connection is in schema-less state. Preparing the same statement would retrieve the one from above, which is prepared for S1


Yes, nice catch!!

I think the solution for this is to initialise the currentSchema & currentCatalog like:

```java
    this.currentSchema = originalSchema != null ? originalSchema : "@default";
    this.currentCatalog = originalCatalog != null ? originalCatalog : "@default";
```

So here `@default` is used instead of `null` in the key for prepared statements.  We can safely use `@default` because it is not a valid sql identifier as it starts with `@`.

> SQL identifiers and key words **_must begin with a letter (a-z, but also letters with diacritical marks and non-Latin letters) or an underscore (_)_**. Subsequent characters ...

Reply:
@rPraml ... just seeing if you can review the last commit (which initialises `currentSchema` using `"@default"` instead of `null`. Cheers, Rob.

Reply:
Will review tomorrw...